dff/stats/instrumentor.py

```diff
@@ -0,0 +1,193 @@
+    e.g. enable and disable logging at runtime.
+
+    .. code-block::
+
+        dff_instrumentor = OtelInstrumentor()
+        dff_instrumentor.instrument()
+        dff_instrumentor.uninstrument()
+
+    Opentelemetry provider instances can be optionally passed to the class constructor.
+    Otherwise, the global logger, tracer and meter providers are leveraged.
+
+    The class implements the :py:meth:`~__call__` method, so that
+    regular functions can be decorated using the class instance.
+
+    .. code-block::
+
+        @dff_instrumentor
+        def function(context, pipeline, runtime_info):
+            ...
+
+    :param logger_provider: Opentelemetry logger provider. Used to construct a logger instance.
+    :param tracer_provider: Opentelemetry tracer provider. Used to construct a tracer instance.
+    :parame meter_provider: Opentelemetry meter provider. Used to construct a meter instance.
+    """
+
+    def __init__(self, logger_provider=None, tracer_provider=None, meter_provider=None) -> None:
+        super().__init__()
+        self._logger_provider: Optional[LoggerProvider] = None
+        self._tracer_provider: Optional[TracerProvider] = None
+        self._meter_provider: Optional[MeterProvider] = None
+        self._logger: Optional[Logger] = None
+        self._tracer: Optional[Tracer] = None
+        self._meter: Optional[Meter] = None
+        self._configure_providers(
+            logger_provider=logger_provider, tracer_provider=tracer_provider, meter_provider=meter_provider
+        )
+
+    def __enter__(self):
+        if not self.is_instrumented_by_opentelemetry:
+            self.instrument()
+        return self
+
+    def __exit__(self):
+        if self.is_instrumented_by_opentelemetry:
+            self.uninstrument()
+
+    @classmethod
+    def from_url(cls, url: str, insecure: bool = True, timeout: Optional[int] = None):
+        """
+        Construct an instrumentor instance using only the url of the OTLP Collector.
+        Inherently modifies the global provider instances adding an export destination
+        for the target url.
+
+        .. code-block::
+
+            instrumentor = OtelInstrumentor.from_url("grpc://localhost:4317")
+
+        :param url: Url of the running Otel Collector server. Due to limited support of HTTP protocol
+            by the Opentelemetry Python extension, GRPC protocol is preferred.
+        :param insecure: ...
+        :param timeout: Connection timeout in seconds, optional.
+        """
+        set_logger_destination(OTLPLogExporter(endpoint=url, insecure=insecure, timeout=timeout))
+        set_tracer_destination(OTLPSpanExporter(endpoint=url, insecure=insecure, timeout=timeout))
+        set_meter_destination(OTLPMetricExporter(endpoint=url, insecure=insecure, timeout=timeout))
+        return cls()
+
+    def instrumentation_dependencies(self) -> Collection[str]:
+        """
+        :meta private:
```

Comment:
Maybe some dicumentation is still required?

Reply:
Done